### PR TITLE
189 - handle location permission request

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

This PR handles the case when the user clicks "Deny" when asked to grant location permission.  

Ticket: 
https://github.com/medic/cht-android/issues/189

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by latin-panda** [comment]
Hi @jkuester, can you please review this code? Thanks!

**Review by jkuester — COMMENTED** [review]
Sweet, the code looks good!  I just want to make sure I understand exactly what we are changing here.  From what I can tell, we were not considering "COARSE_LOCATION" before and now we will actually include that permission when asking the user.  Is that accurate?  

Also, what are your thoughts on unit tests for this permissions code?  Seems like it would be nice to have something, but I have not spent too much time trying to figure out how hard they would be to setup for this code...

**Comment by latin-panda** [comment]
Hi @jkuester 

> we were not considering "COARSE_LOCATION" before and now we will actually include that permission when asking the user. Is that accurate?

Yes, specially once we move to Android 12 (version 1 of cht-android), some more information [here](https://github.com/medic/cht-android/issues/207).

> unit tests for this permissions code

I haven't figured out how to mock the `EmbeddedBrowserActivity`, I always end up with a Null Pointer Exception for a deep inner property that isn't set. However I've added a couple of unit test that didn't give issues. Wondering if you have any suggestion?  

Thanks!

**Comment by jkuester** [comment]
Okay @latin-panda, so I think your NPE is being caused because the `EmbeddedBrowserActivity` is actually not fully initialized when you just call `new EmbeddedBrowserActivity()`. The object is created, but the activity has not had the initialization magic run by the Android system code.  That means that when the activity tries to create the new intent and it checks the current context for the activity, it blows up since there is no context setup for the activity.  

After some searching it seems like the [`ActivityScenarioRule`](https://developer.android.com/reference/androidx/test/ext/junit/rules/ActivityScenarioRule) is the recommended way to actually initialize those activities for testing.  

Have a look at the sample test that I threw together in [this commit](https://github.com/medic/cht-android/commit/3b6d1f996ce1bd739f69c2a3133f97e8faf4649c).  

Couple of notes about the sample:
- We should probably switch the other tests in that class to use the activity from the rule instead of calling `new EmbeddedBrowserActivity()`
- I had to switch to `3.5.0-alpha03` of the espresso libraries since version `3.4.0` caused a [weird interaction](https://github.com/robolectric/robolectric/issues/6593) between espresso and robolectric....

**Comment by latin-panda** [comment]
@jkuester Thanks for the help, it simplified the unit test and works very well! 
Can you please have a look? I added a couple of null checks

**Review by jkuester — APPROVED** [review]

File: src/test/java/org/medicmobile/webapp/mobile/EmbeddedBrowserActivityTest.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			) {
+			contextCompatMock.when(() -> ContextCompat.checkSelfPermission(any(), eq(ACCESS_FINE_LOCATION))).thenReturn(PERMISSION_GRANTED);
+			contextCompatMock.when(() -> ContextCompat.checkSelfPermission(any(), eq(ACCESS_COARSE_LOCATION))).thenReturn(PERMISSION_GRANTED);
+
+			scenarioRule
+				.getScenario()
+				.onActivity(embeddedBrowserActivity -> {
+					assertTrue(embeddedBrowserActivity.getLocationPermissions());
+
+					medicLogMock.verify(() -> MedicLog.trace(
```
[the remark this answers is not in the document]
Comment (latin-panda) [reply src/test/java/org/medicmobile/webapp/mobile/EmbeddedBrowserActivityTest.java:68]: Done!

**Pull request merged**